File: 47.permutations-ii.py
#
# @lc app=leetcode id=47 lang=python3
#
# [47] Permutations II
#

# @lc code=start
<<<<<<< HEAD
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        
# @lc code=end

=======
from typing import *
import copy
class Solution:
    def __init__(self):
        self.out = []
        self.res = []
        self.visited = None

    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        self.visited = [False] * len(nums)
        self.helper(nums)
        return self.res

    def helper(self,nums):
        if len(self.out) == len(nums):
            self.res.append(copy.copy(self.out))
            return
        for index,num in enumerate(nums):   
            if self.visited[index]:
                continue
            if index > 0 and num == nums[index-1] and not self.visited[index-1]:
                continue
            self.out.append(num)
            self.visited[index] = True
            self.helper(nums)
            self.out.pop()
            self.visited[index] = False


# @lc code=end


if __name__ == "__main__":
    print(Solution().permuteUnique([1,1,2]))

>>>>>>> RFC:change mac

File: golang/47.permutations-ii.go
/*
 * @lc app=leetcode id=47 lang=golang
 *
 * [47] Permutations II
 */
package main

import "sort"

// @lc code=start
func permuteUnique(nums []int) [][]int {
	sort.IntSlice(nums).Sort()

	used := make([]bool, len(nums))
	current := []int{}
	result := [][]int{}
	helper47(nums, current, used, &result)
	return result
}

// 副本中的array指针与原slice指向同一个地址，所以当修改副本slice的元素时，原slice的元素值也会被修改。但是如果修改的是副本slice的len和cap时，原slice的len和cap仍保持不变
// 如果在操作副本时由于扩容操作导致重新分配了副本slice的array内存地址，那么之后对副本slice的操作则完全无法影响到原slice，包括slice中的元素

func helper47(nums []int, current []int, used []bool, results *[][]int) {
	if len(current) == len(nums) {
		tmp := []int{}
		for _, n := range current {
			tmp = append(tmp, n)
		}
		*results = append(*results, tmp)
		return
	}

	for idx, num := range nums {
		if idx > 0 && nums[idx] == nums[idx-1] && !used[idx-1] {
			continue
		}
		if !used[idx] {
			current = append(current, num)
			used[idx] = true
			helper47(nums, current, used, results)
			current = current[0 : len(current)-1]
			used[idx] = false
		}
	}
}

// @lc code=end
